src/tcp_connections.rs

```diff
@@ -76,6 +78,36 @@ pub fn listen(port: u16) -> IoResult<(Receiver<(TcpStream, SocketAddr)>, TcpList
 
     let tcp_listener2 = try!(tcp_listener.try_clone());
     let _ = thread::Builder::new().name("TCP listener".to_string()).spawn(move || {
+        for ifa in getifaddrs::getifaddrs() {
+            let ifa = match ifa.addr {
+                IpAddr::V4(a) => a,
+                _ => continue,
+            };
+            match tcp_listener2.local_addr() {
+                Ok(addr) => {
+                    match addr {
+                        SocketAddr::V4(addr) => {
+                            let port = addr.port();
+                            let _ = thread::Builder::new()
+                                .name("IGD".to_string()).spawn(move || {
+                                match igd::search_gateway_from(ifa) {
+                                    Ok(gateway) => {
+                                        let _ = gateway
+                                            .add_port(igd::PortMappingProtocol::TCP,
+                                                      port,
+                                                      SocketAddrV4::new(ifa, port),
+                                                      0, "crust");
+                                    },
```

Comment:
we are missing ability to know our_external_endpoint (call to `gateway::get_external_ip()`). This information needs to be stored in connection manager and should be provided on call to  `connection_manager::get_own_endpoints() -> Vec<Endpoint>` (Returning external endpoint (if IGD Protocol supported) along with all its own listening endpoint (all interfaces)).